Add query for product filters

roykho opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

@imanish003  @kmanijak - roping y'all in this PR because you have the most knowledge in this code space.

This PR is extraction of a larger PR here https://github.com/woocommerce/woocommerce/pull/52516

The reason for separating it is because some of the code changes touches on publicly available implementation. The goal of the larger PR is to add additional statuses to the Experimental Product Filters block.

Here are some high level notes for this PR:

- Added query clauses and logic for On Sale product statuses.
- Added query parameter `filter_status`. This is to eventually facilitate the combination of stock statuses and product statuses like On Sale. So the new Product Filters block will process `https://site.com/?filter_status=instock,onsale` etc.

We're mainly just making sure there are no regressions to the existing public product filters block.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to Appearance -> Editor -> Template -> Product Catalog
2. Add the Product Filters Pattern
3. Ensure all filters are function as expected

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Add onsale and filter_status clauses and queries
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

roykho committed:
Add query for product filters

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterStatus.php:
This shouldn't be included in this PR right?

roykho replied:
It needs it because we're calling its class constant here https://github.com/woocommerce/woocommerce/pull/52841/files#diff-933825fdae3edb1301562e70acef6627856dd1c39c7475056ef21ac4f0ee9941R812

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductQuery.php:
I don't think we need to update this block, as we deprecated it in favor of Product Collection.

dinhtungdu replied:
Instead, we should update the PC block to support the new param.

roykho replied:
I removed the changes to the deprecated block https://github.com/woocommerce/woocommerce/pull/52841/commits/108bd77df5369a13d7b420d5d3946fe7dcb20356

Added the same changes to the PC block https://github.com/woocommerce/woocommerce/pull/52841/commits/60aea4797f609aa2b01ca60c3c7c66306f659176

roykho committed:
Remove changes to deprecated block

roykho committed:
Add query vars for filter blocks into PC

roykho committed:
Fix merge conflicts

roykho committed:
Fix linting

roykho committed:
Change only what is necesary for this PR

roykho committed:
Merge branch 'trunk' into add-query-for-product-filters

kmanijak commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/QueryBuilder.php:
```diff
@@ -121,6 +123,26 @@ private function get_valid_query_vars() {
+			'status_filter_query_args' => array( ProductFilterStatus::FILTER_STATUS_QUERY_VAR ),
+		);
+
+		return array_reduce(
+			array_values( $query_vars ),
+			function ( $acc, $query_vars_filter_block ) {
+				return array_merge( $query_vars_filter_block, $acc );
+			},
+			$public_query_vars
+		);
```
If I understand it correctly, this reduce is equivalent to:

```
foreach ($query_vars as $query_vars_filter_block) {
    $public_query_vars = array_merge($query_vars_filter_block, $public_query_vars);
}
return $public_query_vars;
```

It may be just me, but this reduce looks more complex than it should and replacing with `foreach` would make it easier to read 😅

dinhtungdu replied:
I agree with Karol.

Btw, I have a question, why does we treat the new filter different from the the current ones, like we're doing in `get_queries_by_applied_filters()`

roykho replied:
Fixed https://github.com/woocommerce/woocommerce/pull/52841/commits/2835ef2d9dff249346f854a110ae828128561cf4

kmanijak reviewed (commented):
I tested the changes and didn't encounter any problems with current filters 👍 I'd let @dinhtungdu or someone from filters verify the functionality of new filter!

dinhtungdu commented on plugins/woocommerce/src/Blocks/QueryFilters.php:
```diff
@@ -120,6 +131,53 @@ public function get_stock_status_counts( $query_vars ) {
+
+		$onsale_status_counts = array(
+			'onsale' => $result,
+		);
```
I'm curious why you return an array with a single property here instead of just returning the result directly.

dinhtungdu reviewed (changes requested):
The new URL param is working as expected for product catalog template. But it doesn't work for Product Collection blocks that don't sync with current template, while current params do.

Steps to reproduce:
1. On a new page, add Product Collection block > choose Create your own option.
2. Save and view the page on the frontend.
3. Manually add the URL param to the URL: `?filter_status=onsale`.
4. See the product results unchange.
5. Try a different URL, `?filter_color=blue`.
6. See the products are filtered.

Example store:
- https://briefly-realistic-hammerhead.jurassic.ninja/sample-page/?filter_color=blue
- https://briefly-realistic-hammerhead.jurassic.ninja/sample-page/?filter_status=onsale

For the shop page, the query is handled by main query, so the filter added in `QueryFilters` is sufficient to filter the main query. But for Product Collection blocks that don't sync with main query, the block handle the query itself, so the change mades to PC in this PR doesn't seem to work as expected.

roykho committed:
Simplify reduce logic

roykho commented:
Thanks for the review @dinhtungdu 

> For the shop page, the query is handled by main query, so the filter added in QueryFilters is sufficient to filter the main query. But for Product Collection blocks that don't sync with main query, the block handle the query itself, so the change mades to PC in this PR doesn't seem to work as expected.

Yes I didn't think we need this feature on beta launch. 

> I'm curious why you return an array with a single property here instead of just returning the result directly.

Because `get_stock_status_counts` is also keyed. So `get_onsale_status_counts` should follow the same pattern. Ref: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Blocks/QueryFilters.php#L92

roykho committed:
Fix linting

roykho committed:
Merge branch 'trunk' into add-query-for-product-filters

roykho committed:
Merge branch 'trunk' into add-query-for-product-filters

dinhtungdu commented on plugins/woocommerce/src/StoreApi/Schemas/V1/ProductCollectionDataSchema.php:
```diff
@@ -118,6 +118,29 @@ public function get_properties() {
 				],
 			],
+			'onsale_status_counts' => [
+				'description' => __( 'Returns number of products with each onsale status.', 'woocommerce' ),
```
We only have one on sale status, right?

roykho replied:
Updated via https://github.com/woocommerce/woocommerce/pull/52841/commits/4751b2a3ec97abfb8cbba33280586b93ac77a603

dinhtungdu reviewed (commented):
Thanks for the update. I've just realized that we should also update the docs for the Product Collection Data endpoint at https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/StoreApi/docs/product-collection-data.md.

kmanijak commented:
> > For the shop page, the query is handled by main query, so the filter added in QueryFilters is sufficient to filter the main query. But for Product Collection blocks that don't sync with main query, the block handle the query itself, so the change mades to PC in this PR doesn't seem to work as expected.

> > Yes I didn't think we need this feature on beta launch.

> The problem is that people are already using filter blocks with collections that don't sync with main query. So this will create a mixed experience for them. @kmanijak can you confirm?

Current filters - yes, I can confirm some users use them in such conditions. But at the same time, there's no UI that would allow filtering by `?filter_status=onsale` except for introducing filtering via `queryParam`. In that case, I don't see a problem introducing this filtering option in steps (kind of as progressive enhancement):
- step 1: for Product Collection that syncs with query,
- step 2: for Product Collection that doesn't sync with query.
Of course, as long as new filters that expose this filter in UI will work in both cases.

roykho committed:
Update counts to be singular as we're getting only one type of status

roykho commented:
@dinhtungdu 

> The problem I have with it is, it's only a single value. Isn't it clearer to return just the count as int, and possibly change the function name to get_onsale_status_count? I agree on following conventions when possible, but in this case, it doesn't make much sense to me.

Updated via https://github.com/woocommerce/woocommerce/pull/52841/commits/4751b2a3ec97abfb8cbba33280586b93ac77a603

> Thanks for the update. I've just realized that we should also update the docs for the Product Collection Data endpoint at

Do we want to expose the parameter now? Or should we wait until we actually launch the Product filters beta?

roykho committed:
Merge branch 'trunk' into add-query-for-product-filters

roykho committed:
Fix linting

roykho committed:
Remove set query vars for future PR

dinhtungdu commented:
> Do we want to expose the parameter now? Or should we wait until we actually launch the Product filters beta?

The moment we merge it into trunk, it's already publicly available. The param doesn't have to sync with Product Filters, it's part of Store API.

So, IMO, we should add documentation for the new parameter, which will help Store API users. Probably also note the expected release that contains the new param so users with the current version are not confused. We can do it in a follow-up PR too, if you prefer.

This is looking good to me. The only missing piece is some unit tests for the new parameter. We already have tests for the endpoint [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/tests/php/src/Blocks/StoreApi/Routes/ProductCollectionData.php). Can you please add some? I'm sorry for the incremental reviews, I didn't notice the missing tests previously.

roykho committed:
Clean up and simplify return response

roykho committed:
Add unit tests

roykho committed:
Fix tests

roykho committed:
Fix linting

roykho commented:
Hi @dinhtungdu - thanks for the review so far. I made some changes.

- I renamed the `onsale_status_count` to be just `onsale_count` as it is more fitting for the route. `status` is only really used in the context of the product filters.
- Added unit tests
- Updated the remaining `onsale_counts` to `onsale_count` as we're only counting one number.
- Added endpoint documentation

roykho committed:
Update key name

roykho commented:
With the structural changes we did with product filters, this PR is stale. Closing this for now.
